teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+only used in the context of writing Tasks results, there will be no conflict - see
+[variable replacement with object params](#variable-replacement-with-object-params) for more detail).
+
+But still, features we add in the future may conflict, for example if we added an optional params feature, we might
+provide a variable replacement like `$params.foo.bound`. What if the `foo` param was an object with a key called `bound` ?
+
+For example:
+
+```yaml
+results:
+  - name: foo
+    type: object
+    properties:
+      bound: {
+        type: string
+      }
+```
+
+[TEP-0080](https://github.com/tektoncd/community/blob/main/teps/0080-support-domainscoped-parameterresult-names.md#proposal)
+added support for using `[]` syntax instead of `.` to support variables which are have names that include `.`. Resolving
+the ambiguity here would only require using `$params.foo.bound` to refer to the built in variable replacement, and
+`$params.foo["bound"]` to unambiguously refer to the key `bound` within `foo`. Other options include:
+
+1. Do not allow keys to be defined which conflict with variable replacement
+    * Con: this means that a Task which is perfectly fine today might break tomorrow if we introduce new variable
+      replacement - in the above example, `foo.bound` is perfectly valid today and would be allowed but would suddenly
+      stop working after the `bound` feature is introduced
+2. Let the defined object override the variable replacement; i.e. in this case the built in replacement will not be
+   available
+
+## Test Plan
+
+In addition to unit tests the implementation would include:
+
+* At least one reconciler level integration test which includes an object param and an object result
+* A tested example of a Task, used in a Pipeline, which:
+    * Declares an object param
+    * Emits an object result
+    * In the Pipeline, another Task consumes a specific value from the object
+
+## Design Evaluation
+
+[See also the TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) design evaluation.
+
+* [Reusability](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability):
+    * Pro: PipelineResources provided a way to make it clear what artifacts a Pipeline/Task is acting on, but
+      [using them made a Task less resuable](https://github.com/tektoncd/pipeline/blob/main/docs/resources.md#why-arent-pipelineresources-in-beta)
+      - this proposal introduces an alternative that does not involve making a Task less reusable (especially once
+      coupled
+      with [TEP-0044](https://github.com/tektoncd/community/blob/main/teps/0044-decouple-task-composition-from-scheduling.md))
+* [Simplicity](https://github.com/tektoncd/community/blob/main/design-principles.md#simplicity)
+    * Pro: Support for emitting object results builds on [TEP-0076](https://github.com/tektoncd/community/pull/477)
```

Comment:
This is maybe more about flexibility, however for a user to be able to pass a json dict as a result they'll have to define a schema for the result and the same schema for input params that consume that result (in any).
I think this is perfectly acceptable in cases where users want to have extra validation and observability into the pipelines, but it might be too much for the more casual author. I wonder if we should support schema-less as well?

Reply:
interesting idea but I'm not 100% sure what this would look like, could you provide an example?